client/network-gossip/src/bridge.rs

```diff
@@ -0,0 +1,254 @@
+							remote,
+							messages.into_iter()
+								.filter_map(|(proto, data)| if proto == proto_name {
+									Some(ConsensusMessage { engine_id, data })
+								} else { None })
+								.collect()
+						);
+					},
+					Event::Dht(_) => {}
+				}
+			}
+		});
+
+		gossip_engine
+	}
+
+	/// Closes all notification streams.
+	pub fn abort(&self) {
+		self.inner.lock().state_machine.abort();
+	}
+
+	pub fn report(&self, who: PeerId, reputation: i32) {
+		self.inner.lock().context.report_peer(who, reputation);
+	}
+
+	/// Registers a message without propagating it to any peers. The message
+	/// becomes available to new peers or when the service is asked to GossipEngine
+	/// the message's topic. No validation is performed on the message, if the
+	/// message is already expired it should be dropped on the next garbage
+	/// collection.
+	pub fn register_gossip_message(
+		&self,
+		topic: B::Hash,
+		message: Vec<u8>,
+	) {
+		let message = ConsensusMessage {
+			engine_id: self.engine_id,
+			data: message,
+		};
+
+		self.inner.lock().state_machine.register_message(topic, message);
+	}
+
+	/// Broadcast all messages with given topic.
+	pub fn broadcast_topic(&self, topic: B::Hash, force: bool) {
+		let mut inner = self.inner.lock();
+		let inner = &mut *inner;
+		inner.state_machine.broadcast_topic(&mut *inner.context, topic, force);
+	}
+
+	/// Get data of valid, incoming messages for a topic (but might have expired meanwhile)
+	pub fn messages_for(&self, topic: B::Hash)
+		-> mpsc::UnboundedReceiver<TopicNotification>
+	{
+		self.inner.lock().state_machine.messages_for(self.engine_id, topic)
+	}
+
+	/// Send all messages with given topic to a peer.
+	pub fn send_topic(
+		&self,
+		who: &PeerId,
+		topic: B::Hash,
+		force: bool
+	) {
+		let mut inner = self.inner.lock();
+		let inner = &mut *inner;
+		inner.state_machine.send_topic(&mut *inner.context, who, topic, self.engine_id, force)
+	}
+
+	/// Multicast a message to all peers.
+	pub fn gossip_message(
+		&self,
+		topic: B::Hash,
+		message: Vec<u8>,
+		force: bool,
+	) {
+		let message = ConsensusMessage {
+			engine_id: self.engine_id,
+			data: message,
+		};
+
+		let mut inner = self.inner.lock();
+		let inner = &mut *inner;
+		inner.state_machine.multicast(&mut *inner.context, topic, message, force)
+	}
+
+	/// Send addressed message to the given peers. The message is not kept or multicast
+	/// later on.
+	pub fn send_message(&self, who: Vec<network::PeerId>, data: Vec<u8>) {
+		let mut inner = self.inner.lock();
+		let inner = &mut *inner;
+
+		for who in &who {
+			inner.state_machine.send_message(&mut *inner.context, who, ConsensusMessage {
+				engine_id: self.engine_id,
+				data: data.clone(),
+			});
+		}
+	}
+
+	pub fn announce(&self, block: B::Hash, associated_data: Vec<u8>) {
+		// TODO: self.announce_block(block, associated_data)
+	}
+
+	pub fn set_sync_fork_request(&self, peers: Vec<network::PeerId>, hash: B::Hash, number: NumberFor<B>) {
```

Comment:
While I know it is a convenient place to put this, this is not related to gossiping but much rather a way to communicate with another component (sync). I am guessing that only the Grandpa module needs this function. Thus this could e.g. be a channel between the Grandpa and the Sync instance. What do you think?

Reply:
There's actually already an abstraction for that, but it's used only in one module: https://github.com/paritytech/substrate/blob/2fa6a2647ee430604139c2b9efb2e156983d166f/client/finality-grandpa/src/lib.rs#L271

Reply:
Right. This is used as a way to do dependency injection and was meant for the finality grandpa module only.

Are you suggesting to have `NetworkService` implement this trait and thus still passing a clone of `NetworkService` into grandpa?